File: 周赛/156/no3.py
from mytools import *
class Solution(object):
    def maxWeight(self, n, edges, k, t):
        """
        :type n: int
        :type edges: List[List[int]]
        :type k: int
        :type t: int
        :rtype: int
        """
        g = [[] for _ in range(n)]
        for u, v, w in edges:
            g[u].append((v, w))
        res = [-1]
        def dfs(e, w_sum, u):
            if e == k:
                res[0] = max(res[0], w_sum)
            for v, w in g[u]:
                if w_sum + w < t:
                    w_sum += w
                    e += 1
                    dfs(e, w_sum, v)
                    e -= 1
                    w_sum -= w
        for i in range(n):
            dfs(0, 0, i)
        return res[0]
        

File: 周赛/157/no3.cpp
#include "tools.h"
class Solution
{
public:
    using ll = long long;
    const int MOD = 1e9 + 7;
    ll mypow(ll a, ll e)
    {
        ll r = 1;
        while (e > 0)
        {
            if (e & 1)
                r = (r * a) % MOD;
            a = (a * a) % MOD;
            e >>= 1;
        }
        return r;
    }

    int assignEdgeWeights(vector<vector<int>> &edges)
    {
        int n = edges.size() + 1;
        vector<vector<int>> g(n + 1);
        for (auto &e : edges)
        {
            int u = e[0], v = e[1];
            g[u].push_back(v);
            g[v].push_back(u);
        }

        int L = 0;
        auto dfs = [&](auto &&self, int u, int parent, int d) -> void
        {
            L = max(L, d);
            for (int v : g[u])
            {
                if (v == parent)
                    continue;
                self(self, v, u, d + 1);
            }
        };
        dfs(dfs, 1, 0, 0);
        return (int)mypow(2, L - 1);
    }
};
